docs/website/docs/extensions/index.md

```diff
@@ -0,0 +1,442 @@
+* Device code may be running out-of-process (sandbox/enclave) and the library
+  functions must be available where the dispatches run and not where they are
+  launched (such as being linked into the sandbox binary, if separate from the
+  main process binary).
+* The stack must be used to pass arguments/results to external calls via a
+  single pointer and there is no libffi-like functionality for magically calling
+  arbitrary C functions. Users must provide the shims they need.
+* Thread-local storage is unavailable in the called code (it may be usable, but
+  it is not guaranteed it'll work on all platforms and leaks are likely).
+* No heap allocator is provided and the use of libc malloc is unsupported.
+
+Most of the constraints here come from the [SPMD](https://en.wikipedia.org/wiki/SPMD)
+parallelism model, platform-agnostic deployment format, and overall
+data-oriented design of IREE. Code operating in this fashion has a certain shape
+and that is usually not the same as big legacy single-threaded CPU-focused BLAS
+libraries that perform their own caching, internal thread and state management,
+and other shenanigans. IREE is not designed to wrap such things and if any of
+these notes are issues it is more an indicator that the approach needs
+adjustment than anything else. Trying to bypass or workaround the constraints is
+possible - after all IREE is an open source project and any user is welcome to
+fork it - but unsupported by the core IREE team.
+
+### Pros
+
+* Function resolution at runtime is orthogonal to compiler target specification.
+* Machine code can be shared between the application and IREE artifacts.
+
+### Cons
+
+* IREE compiler and runtime must both be modified.
+* Deeper integration with the IREE codegen compiler infrastructure required.
+* ABI versioning complexity between compiler and runtime.
+* Runtimes must ship the imports for the lifetime of any artifact compiled to
+  use them.
+  * Humans are bad at predicting the future.
+  * Using the same artifact in different binaries at runtime requires changes to
+    each binary - including those that may not be owned by the person producing
+    the artifact.
+  * Weak imports and conditional usage can help but still leads to bloat.
+
+### When to use
+
+* :heavy_check_mark: Calling into opaque closed-source BLAS-like microkernel
+  libraries.
+* :x: Any other cases covered above can be used, especially microkernels that
+  can be represented in MLIR or as statically linked libraries.
+
+### Implementation
+
+The compiler is changed to produce calls to imports via a dynamic import table
+provided to each dispatch function. The import table is declared in the
+executable library for use at runtime. Runtime applications register an import
+provider to resolve named symbols in the import table to C functions that
+marshal arguments and results.
+
+The compiler-side needs some additional work but an example is included here: https://github.com/google/iree/issues/7504.
```

Comment:
nit: mkdocs does not make a hyperlink out of loose URLs :( you'll need to wrap in `[description](url)`